include/swift/SILOptimizer/Utils/ScopeOptUtils.h

```diff
@@ -0,0 +1,284 @@
+//===--- ScopeOptUtils.h --------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// A simple scope like construct for use in the SILOptimizer.
+///
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SILOPTIMIZER_UTILS_SCOPEOPTUTILS_H
+#define SWIFT_SILOPTIMIZER_UTILS_SCOPEOPTUTILS_H
+
+#include "swift/Basic/LLVM.h"
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SIL/SILValue.h"
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/SmallVector.h"
+#include <functional>
+
+namespace swift {
+namespace siloptimizer {
+
+template <typename Result, typename... Args>
+struct Cleanup {
+  std::function<Result(Args...)> func;
+};
+
+/// A simple stack data structure that manages a stack of std::function cleanups
+/// that can be invalidated by index.
+///
+/// It is called asserting, since we force our user to explicitly pop the scope
+/// and abort otherwise.
+template <typename Result, typename... Args>
+class AssertingScope {
+  using ScopeCleanup = Cleanup<Result, Args...>;
+  SmallVector<Optional<ScopeCleanup>, 64> cleanups;
+  bool didPop = false;
+
+protected:
+  ~AssertingScope() { assert(didPop && "Did not pop scope?!"); }
+
+  AssertingScope() {}
+
+  /// Push a new SIL function based cleanup. Meant to be very
+  /// lightweight. Returns the index of the cleanup.
+  unsigned pushScope(std::function<Result(Args...)> func) {
+    assert(!didPop && "Can not reuse scope once used?!");
+    unsigned index = cleanups.size();
+    cleanups.emplace_back(ScopeCleanup{func});
+    return index;
+  }
+
+  /// Invalidate the cleanup at the specific index. Asserts on index too big and
+  /// on Optional has value.
+  void invalidateCleanup(unsigned cleanupIndex) {
+    assert(!didPop && "Should not invalidate once popped?!");
+    assert(cleanupIndex < cleanups.size());
+    assert(cleanups[cleanupIndex].hasValue());
+    cleanups[cleanupIndex] = None;
+  }
+
+  /// Pop the scope, running all non-Optional cleanups, and setting didPop. Once
+  /// this is done the scope has been consumed and no longer has any state.
+  ///
+  /// Can only be done once the scope has been moved.
+  void pop(Args... args) && {
+    while (cleanups.size()) {
+      auto val = cleanups.pop_back_val();
+      if (val) {
+        val->func(std::forward<Args>(args)...);
+      }
+    }
+    didPop = true;
+  }
+
+  /// Can not be called once the scope is moved.
+  void nonDestructivePop(Args... args) const {
```

Comment:
For future consideration. It seems like the point of `nonDestructivePops` would be to copy the scope without making a physical copy. So it would be more robust/usable to allow scopes to be either copied or chained with multiple scopes sharing the cleanup vector. Then popping is only destructive when you're the original or unique owner of those cleanups. I don't know if copying (with CoW behavior) or chaining with shared cleanups would make more sense in practice.

Reply:
I would need to think about this. The point of this is to enable one to along early exiting paths pop all current scoped things.

In SILGen, the scope data structure allows for a shared cleanup vector. I think that would be a great add-on. I was trying to keep this simple.